### ISSUE TITLE
WTI server should fatal-abort when scoreboard login fails

### ISSUE THREAD

**Issue by clevengr** [pc2ccs/pc2v9#117]
**Describe the issue**: 
  When the WTI-API server is started (using the **pc2wti** command), one of the things it must do is login to the PC2 server with a "scoreboard" account.  (The acct name/pw for this login comes from the WTI's **pc2v9.ini** file; if no credentials are present in that file then it uses default acct/pw values of _scoreboard2/scoreboard2_.)

(Note: this behavior -- logging into a scoreboard account -- is only implemented in the WTI version represented by PR #116; prior to that code there was no implementation of "scoreboard" in the WTI.)

  If the scoreboard login fails, this should be a fatal error condition.  Currently however it is not; rather, it simply throws an exception on the WTI console and continues (the login failure is also logged in the PC2 server log under the running "profile").

  Furthermore, this exception does not even happen at WTI-API startup.  Rather, it doesn't occur until the first reference to some endpoint in the WTI Jetty/Jersey server.  This means the WTI server will appear to start up fine, then at some later point in time will fail (for example, when the first team logs in) -- which is bad since this might happen a long while after the WTI server was started.

  The WTI-API server needs to abort immediately on startup if it is not possible to use the provided (or default) credentials to log in to the PC2 server as a scoreboard client.

**To Reproduce**:
1. Download a version of pc2 which has PR #116 code in it (if PR #116 has already been merged into PC2V9, then any "develop" version after the merge will work; otherwise, download the code directly from PR #116).
1. Start a pc2 server using the dowloaded pc2v9 system,  with a contest configured with at least one team account (for example, use **--load sumithello** to load the _sumithello_ sample contest).
1. Go into the WebTeamInterface project folder and edit the **pc2v9.ini** file:  change either the **wtiscoreboardaccount=** property or the **wtiscoreboardpassword=** property (or both) so that they represent an acct/pw combination that will NOT be accepted by the pc2 server.  (Note: if the **sumithello** sample contest is used, this editing can be omitted since the default account/pw is _scoreboard2/scoreboard2_ but the **sumithello** contest is not by default configured with such an account.)
1. Start the WTI-API server using the command "pc2wti".  

**Expected behavior**
This should cause a fatal abort of the WTI-API server because it has invalid scoreboard login credentials.

**Actual behavior**:
This will display a series of successful startup messages on the console.

Continuing:
1. Open a browser and enter the address "http://localhost:8080".  This should produce the WTI login screen.
1. Enter valid account/pw credentials for a team.

This will successfully log in the team, but will throw a large number of exceptions on the WTI console (whereas what it SHOULD do is **abort** the WTI-API server).

**Environment**: 
This was tested (reproduced) under Windows 10 with Java 8_201 and Chrome Version 83.0.4103.116, but is almost assuredly reproducible in virtually any environment/platform.

**Log Info**:  
The following shows the exceptions on the WTI console:

[WTI.console.scoreboardLoginFailure.exceptions.txt](https://github.com/pc2ccs/pc2v9/files/4855099/WTI.console.scoreboardLoginFailure.exceptions.txt)

**Additional context**: 
  When WTI is first started, it starts a Jetty server and configures Jersey (the JAX-RS handler) with classes to handle REST endpoints.  Among the classes configured by name into Jersey (in `/WTI-API/src/main/config/WebServer.java`, method `getJerseyHandler()`) are WTI-API classes `controllers.TeamsController` and `controllers.ContestController`.  

It is the latter class (`ContestController`) which performs the 'scoreboard login' -- so this is where the fatal error should occur.  In fact, `ContestController` has a _static initialization block_ which throws a `RuntimeException` on a PC2 scoreboard login failure.

Note that the above log file clearly shows that this RuntimeException is being thrown -- just not when the WTI first starts up (rather, it is thrown when a team logs in -- presumably because this (indirectly) causes instantiation of a `ContestController`.)

Note that these classes are only loaded into Jersey by _name_ (that is, they are not actually _instantiated_ at this point).  This means that the static initialization block (and hence the exception condition) does not occur during WTI startup, but only later when something actually requires Jersey to _instantiate_ a `ContestController`.  It appears that this occurs when a team logs in; hence the console exceptions described above upon team login.

Jetty documentation (at https://www.eclipse.org/jetty/javadoc/current/org/eclipse/jetty/servlet/ServletHolder.html) states that Jetty ServletHolders can be forced to be initialized "when the handler is started" by setting an "initOrder" to a value greater than zero.  However, the WTI `WebServer` code already calls `setInitOrder(1)` for the handler, which doesn't change anything.

Several possibilities come to mind: 
1. The version of Jetty/Jersey being used by WTI does not correctly implement the `setInitOrder()` parameter.
2. The "handler" is not actually "started" when the server starts.

A test was made to call the `start()` method in each configured handler after the Jetty server was started, but this had no effect; the WTI still started up and didn't throw the login-failure exception until a team login occurred.

**Comment by clevengr** [comment]
I have spent many, many hours looking into this issue (starting long before the bug/Issue was actually filed).  Recently I came across some StackOverflow discussions (e.g. [this](https://stackoverflow.com/questions/28114602/how-do-i-get-my-jersey-2-endpoints-to-eagerly-initialize-on-startup/28123656#28123656), [this](https://stackoverflow.com/questions/29289245/immediate-annotation-use-in-jersey2), and [this](https://stackoverflow.com/questions/19552111/how-to-load-servlet-on-startup-using-jersey-2-3-with-embedded-jetty)) that imply that the problem is that setInitOrder(1) loads/initializes the _**ServletContainer**_ inside the ServletHolder, but not the classes configured inside that ServletContainer.  I'm continuing to investigate.

**Comment by clevengr** [comment]
I've now spent many MORE hours researching this issue, trying to find a "clean" way to get Jetty/Jersey to instantiate (or at least, load) the ContestController class so that its static initialization block (which tests the PC2 scoreboard login credentials) would be executed at Jetty startup (rather than waiting until a Jetty endpoint involving the ContestController was referenced).  All of my many hours of searching and trying various things have failed.

However, I did find a 'workaround':  I added code to the WebServer class (which is what starts Jetty) to first directly try to login to the PC2 server using the WTI's scoreboard account credentials.  If this login fails, it throws an exception and bypasses the Jetty startup entirely.  This accomplishes the intended goal of causing a fatal abort of Jetty startup if the scoreboard login fails.

I have pushed the code to do this as an update to PR #116, so I'm closing this issue (since it is now resolved in that PR).

I would love it if someone could identify a way to make this happen as part of the normal Jetty/Jersey startup process, but until then I have tested and verified that the above workaround works.
